### ISSUE TITLE
Create test cases to evaulate effectiveness of AL strategy 

### ISSUE THREAD

**Issue by chaitraliagashe** [inception-project/inception#1234]
**Is your feature request related to a problem? Please describe.**
Add a datasplitter for active learning module. Currently there is no way to test effectiveness of AL strategy. In order to do that we need to train recommender with and without AL strategy and then asses the impact of AL strategy. 

**Describe the solution you'd like**
First step towards this goal is to create a AL splitter which would do the following:
1) extract data from an annotated corpus (one or more documents)
2) create separate train and test sets from that data
3) extract an initial set of say the first 10 gold standard data points form the training set (assuming a user who starts annotating from the beginning)
4) train a recommender
5) evaluate the recommender on the test set
6) apply the AL strategy to get the next 10 best data points according to the strategy (assuming the user now enables active learning)
7) add the gold data for these 10 points to the training data
8) go back to 5

**Comment by chaitraliagashe** [comment]
In Order to do this, the first step would be to create a Splitter for AL.
As mentioned in the steps above, we need to add the data-points received from AL strategy to the training set and retrain the recommender. We need to ensure that we consider all the newly added points in the training set. The ALSplitter will take care of this.

**Comment by chaitraliagashe** [comment]
I have a question related to how the recommenders are trained - 
There is a parameter called RecommenderContext - the documentation for this suggests that it is used in order to save data between runs. Does this mean that the recommender can be trained incrementally? That is if I train a recommender with 10 points in the first run, and then further train it with 10 more points, model will be generated using all 20 points? If so, please let me know how that can be done.

**Comment by reckart** [comment]
> That is if I train a recommender with 10 points in the first run, and then further train it with 10 more points, model will be generated using all 20 points? If so, please let me know how that can be done.

Currently, all recommenders are trained from the scratch. We want to be able to support incrementally trained recommenders, but so far there is no recommender which implements this feature.

Currently, the recommenders store their models in the context during training and pick it up from there during prediction (mind that in production prediction can happen quite a bit more often than training). When a new model is trained, the old one is replaced.

But in principle, a recommender that is capable of performing incremental training should be able to pick up a previously trained model from the context and augment it.

**Comment by chaitraliagashe** [comment]
Thank you for the clarification

**Comment by chaitraliagashe** [comment]
I feel we should not put unnecessary intelligence in the Splitter code. All the Splitters so far have minimal intelligence and they split the data into training and test based on some basic criteria. The only problem with using existing splitter is that if we do not increase the training portion in appropriate amount, we may end-up losing datapoints generated by AL strategy. I suggest this could be as simple as traversing the list of data-points in reverse order in the ALSplitter and selecting the percentage of training data appropriately.

**Comment by reckart** [comment]
Why does the order in which the data points are traversed make a difference?

**Comment by chaitraliagashe** [comment]
The current implementation of Splitters accepts the data points in a list. If we use arraylist/linkedlist, insertion order is preserved. When we generate more data-points and add them to the list of original data-points, they will go in the end. If the splitter reads them in revers order and appropriate split point is selected, we can gurantee that we have included points generated by AL strategy.

**Comment by reckart** [comment]
The reason I was thinking about using a splitter here is because the datapoints may be quite unevenly distributed between documents. Splitters can operate at the level of datapoints whereas any other mechanism would probably have to operate at the level of entire documents.

In the current implementation, the  `DataSplitter` interface offers only one method `getTargetSet(object)` which receives a single data point and returns to which set (training, test, ignore) it belongs. 

The `IncrementalSplitter` is currently the most intelligent one and seems to be would be similar in function to an ALSplitter in the sense that is not only a simple splitter but actually an iterator over splitters. 

That said, I agree that the present architecture seems to make it quite annoying to use splitters here:

* `ALSplitter`depend on an `ALStrategy`
* `ALStrategy` requires `List<SuggestionGroup>`
* `List<SuggestionGroup>` needs to be generated from a trained model
* the model is trained internally in the `RecommendationEngine.evaluate(CASes, Splitter)` method
* the only "standardized" way we have of accessing the predictions generated from this model (and to transform them into some `List<SuggestionGroup>` seems to be the information that is being added to the `EvaluationResult` (i.e. currently a `List<LabelPair>`). But we'd have to come up  at least with some `ScoredLabelPair` because AL requires a score.

So assuming we had the ability to obtain scored results from the `EvaluationResult` object, we'd still need to find some suitable way of making that information available to the `ALSplitter`. And of course, we'd only get this information **after the first round**, so in the first round, the splitter would in any case have to use some baseline method to select datapoints (e.g. falling back to a `PercentageBaseSplitter`).

As "pseudo-code", all this taken together might look in a unit test (similar to `IncrementalSplitterTest`) like:

```
            // Use AL strategy and fall back to PercentageBasedSplitter for first round
            ALSplitter splitter = new ALSplitter(alStrategy, new PercentageBasedSplitter(k, 4));
            ...
            EvaluationResult prevResult = null
            while (splitter.hasNext()) {
                splitter.update(prevResult);
                splitter.next();
                result = sut.evaluate(casList, splitter);
            }
```

Does that make any sense to you?

**Comment by chaitraliagashe** [comment]
Let me go through this in detail and think about it. I will get back to you once I understand it clearly. I am a bit busy for next 2 days, so it may take a few days.

**Comment by chaitraliagashe** [comment]
We did the previous refactoring tasks for getting the annotation suggestions - extracting suggestions from the data recommender has produced - https://github.com/inception-project/inception/issues/1033

We faced some hick-ups when getting the changes in for the PR - https://github.com/inception-project/inception/pull/1060. I had two branches open by mistake for the same.

It now appears to me that the changes in the branch - Refactoring/1060 refactor prediction task.java #1065 - are not reflected in the master. They contain the refactoring that had PreditionTask.java refactored into a util class.

I feel once we have this sorted out, we will be able to extract annotation suggestions to be given to ALStrategy. After ALStrategy predicts the points, they should be added to the collection of data points which form part of training set. Now this is where it gets challenging - recommenders accept CAS lists while we have objects of AnnotationSuggestions. I am not sure how to handle this tranformation. 

If we can do this, then we will have an updated list of CAS to perform the evaluate method after ALStrategy.

**Comment by reckart** [comment]
@chaitraliagashe I had [changed](https://github.com/inception-project/inception/pull/1060/commits/3b39691ec2a1adddae4a30268808b23717b427ec) your PR such that the code was moved not into `PredictionUtil` but into `RecommendationServiceImpl`.

**Comment by chaitraliagashe** [comment]
Ok. Found it.
Do you think the rest of the approach makes sense?

**Comment by reckart** [comment]
`RecommendationServiceImpl.extractSuggestions(User, CAS, SourceDocument, Recommender)` returns `List<AnnotationSuggestion>` - does that help?

**Comment by chaitraliagashe** [comment]
I know that extractSuggestions gives List<AnnotationSUggestions>. We did the refactoring task in order to have this method available to - AL module so that we will be able to test it in isolation. 
This solves one problem - making the recommender predictions available to ALStrategy. 
Our proposed approach to test ALStrategy is as follows:

1. Take "n" data-points from the corpus and train a recommender on them.
2. Generate suggestions for the next "m" datapoints using this trained recommender (extractSugegstion method)
3. Give this suggestions to ALStrategy and generate the AL recommended points for each of them. 
4. Add these back to training set, retrain the recommender and then asses the performance.

Now, I understand that the recommenders have evaluate method which does the task of evaluating their performance. However, with the current implementation of evaluate method, we do not have the control over the way we train the recommender and also over the training set. 

In order to achieve there are three options -
1. Have ALSplitter which would update the evaluation result generated by the evaluate method in the recommender - I would not prefer it personally - In this case we are updating the evaluation result object rather than updating the training set.
2. Convert List<AnnotationSuggestions>/ List<AnnotationGroup> to CAS annotations. That way we can add the suggestions to Original training set and we can call evaluate method as is - a simple programmatic manipulation of the list of these annotation in the splitter will make sure that the points generated by AL strategy are selected in the training set - I do no prefer it much - I am not sure if we should convert the data points back to CAS - not sure if it is conceptually/ logically right.
3. Refactor evaluate method to give the test framework more control over the way training and testing occurs. Or have a way to let the user train and test/evaluate the recommender separately - i think this is the cleanest approach.

Please let me know your opinion.

**Comment by reckart** [comment]
For 3. what additional control do we have to give over the way training and testing occurs?

**Comment by chaitraliagashe** [comment]
For 3: instead of calling train and test methods as part of evaluation, it could work with evaluation objects.
I understand this could lead to extensive changes. So if the behavior we need in AL testing is going to be uncommon, we can skip using evaluation method and do the evaluation in test-case.

**Comment by reckart** [comment]
The reason we created `evaluation` as a method separate from `train` and `test` was that every recommender internally extracts its data from the CAS in a different way (e.g. as OpenNLP, POSSample, OpenNLP  DocumentSample, DL4J tensors, etc.). If we wanted to have had a generic train/test API which could be used for evaluation as well, then we would have had to introduce a generic data object to carry the data extracted from the CAS. 

Instead of doing that, we implemented the `Splitter` concept which allows us to control the routing of data extracted from the CAS into the training or test sets (or ignore it) without actually having to know the details of  how a particular recommender represents its data internally.

I understand that you think the splitter is not giving sufficient control over the process. That is why I am asking which additional control (parameters) the evaluation method would need - assuming that in general we want to keep the evaluation method.

**Comment by chaitraliagashe** [comment]
@reckart ,
I was a bit busy far past 2 weeks. I will take a look at the whole picture once again and get back to you by Monday.

**Comment by chaitraliagashe** [comment]
Sorry about the delayed response. @reckart   , I thought about how to have more control over the data in the evaluate method of recommenders. I realized on issue in the solution that we are trying to implement. 
The evaluate method in each of the recommenders is for evaluating the performance of the recommender itself, with standard way of training and testing. 
What we are trying to achieve is testing of effectiveness of AL strategy - recommender is a tool that we use to do this. Thus, conceptually we should leverage recommendation capabilities to do so. Thus we should not use the evaluate method in recommender but we should evaluate performance of the recommender before and after AL. Therefore we should call train and test methods explicitly in the test for AL strategy and have an evaluation method - for AL - which would give evaluation result object for every iteration of train-test phase. 
If this is too obscure, please let us discuss this over call.

**Comment by chaitraliagashe** [comment]
If we want to add some parameters in evaluate methods of recommenders, they will become cumbersome with the responsibility they should not architecturally shoulder.
